---
title: "computing PI"
author: "Cyril Matthey-Doret"
date: "28/03/2018"
output: pdf_document
---

# Basic method

## Pseudocode implementation

Reused vcftools algorithm:

* Bases that are not called (NA) will not be counted in pairs
* Fixed bases that are called will count in pairs and decrease pi

```{r eval=F}
pairs = 0
mismatch = 0

for pos in sequence:
  for allele in pos:
    n_other = pos.freq[all] - pos.freq[allele]
    mismatch += pos.freq[allele] * n_other
  pairs += pos.freq[all] * (pos.freq[all] - 1)
  pi[pos] = mismatch/pairs
```

## Variants

* Computing PI per site or in windows ($\frac{sum(mismatch)}{sum(pairs)}$ in each window)
* Including only SNPs vs all fixed sites and SNPs.
* Weighting windows per total number of alleles (only useful if using a smoother)

# Results

## Note on association mapping
* Recolored SNPs according to proportion of homozygous males only
    + 1 peak stands out: Figure \ref{fig:chr_assoc}
* Checked for GWAS hits on non-anchored contigs again after adding MAF filter
    + No interesting contig: Figure \ref{fig:tig_assoc}

## PI values

* Window size affect results a lot (Figure \ref{fig:win_sizes}.
    + 100bp windows are basically same as per site. 1kbp and 10kbp seem more appropriate (?).
    + Note: Many NA in larger window sizes because smaller window sizes can fit more windows in chromosomes.
* Comparing PI of RADseq hits to direct neighbourhood (Figure \ref{fig:RAD_PI})
    + Comparing PI of RADseq *site* vs window 10kbp around. No significant site (from the association) stands out.
    + Peak of PI on chromosome 3 (Figure \ref{fig:neigh_PI}), but RADseq SNP is not much higher than the neighbourhood.
    + Comparing PI values in neighbourhood of significant RAD SNPs versus upstream and downstream regions did not reveal any SNP that "stands out" from local background (Figure \ref{fig:up_down}). Windows might be too small ?

# Ideas/follow-up
Metric to quantify correlation between p-value and PI: Identify peaks of correlation

```{r chr_assoc, echo=F, fig.cap="\\label{fig:chr_assoc}Manhattan plot of the association mapping for CSD using a minor allele frequency filter at 10%. SNPs are colored according to the proportion of heterozygous males (high = bas) and p-values are corrected using the BH procedure. Only SNPs that were anchored to chromosomes are shown"}
knitr::include_graphics("chrom_manhattan.pdf")

```


```{r tig_assoc, echo=F, fig.cap="\\label{fig:tig_assoc}Manhattan plot of the association mapping for CSD using a minor allele frequency filter at 10%. SNPs are colored according to the proportion of heterozygous males (high = bas) and p-values are corrected using the BH procedure. Only SNPs from unanchored contigs are shown. SNP color denotes appartenance to a contig with significant GWAS hit and background stripes show contig boundaries."}
knitr::include_graphics("unanchored_manhattan.pdf")

```

## Different window sizes

Trying out different windows sizes for computing PI.
```{bash merge_PI, echo=F}
pi_dir='../../../../data/wgs_wild/stats/'
merged_PI="$pi_dir/merged_PI.tsv"

echo -e "Chr\tstart\tend\tPI\twin_size" > "$merged_PI"

for pi in $pi_dir/win_w*
do
  size=$(echo $pi | sed 's/.*_w\([0-9]*\)_t.*/\1/')
  sed "s/$/$(printf '\t')$size/g" "$pi" >> "$merged_PI"
done

```

```{r setup, echo=F}
zoom <- 10^6
around <- 10000
packs <- c("ggplot2", "dplyr", "readr", "viridis", "tidyr")
packs <- sapply(packs, library, character.only=T, quietly=T)
setwd("../../../../")
PI_merged <- read_tsv('data/wgs_wild/stats/merged_PI.tsv', col_types = cols())
RAD <- read_tsv("data/assoc_mapping/case_control/case_control_all.tsv", col_names=T, col_types=cols())
PI_sites<- read_tsv("data/wgs_wild/stats/sites_PI.tsv", col_names=F, col_types=cols())
```

```{r PI_win_size, cache=T, cache.lazy=F, echo=F, message=F, warning=F, fig.cap="\\label{fig:win_sizes}Comparison of PI variation s when computed in different window sizes. The color of hexagonal bins denotes the number of windows whose PI value fall in side the bin at the given genomic range."}

PI_merged <- PI_merged %>%
  group_by(win_size) %>%
  mutate(norm_PI = (PI - mean(PI, na.rm=T))/sd(PI, na.rm=T))

ggplot(data=PI_merged, aes(x=((start+end)/2)/zoom,y=PI)) +
  facet_grid(win_size~Chr, scales='free', space = 'free_x') + 
  stat_bin_hex() +
  stat_smooth(col='red') +
  xlab("Genomic position [Mb]") + ylab("PI nucleotidic diversity")
```

```{r RAD_PI, cache=T, cache.lazy=F, echo=F, fig.cap="\\label{fig:RAD_PI}Ratio of the PI values of RADseq SNPs compared to their local (10kb) neighbourhood, shown as a ratio. The color of each SNP shows whether the -log10 of the BH-corrected p-value from the association mapping for CSD was below 10e-5."}

# RAD association mapping data
colnames(PI_sites)<- c("Chr","BP", "PI")

# Shrinking df before merging to speed up operation
PI_small <- PI_sites %>% filter(BP %in% RAD$BP)
snps <- merge(x = PI_small, 
              y = RAD, 
              by = c("Chr", "BP"), 
              all = F)
snps$around10k <- NA
for(chr in unique(RAD$Chr)){
  PI_chr <- PI_sites[PI_sites$Chr==chr,]
  snps$around10k[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x-around) & 
                                                                     PI_chr$BP < (x+around)], na.rm=T))
}

ggplot(data=snps, aes(x=BP/zoom, y=PI/around10k, col=fisher<5)) + 
  facet_grid(~Chr, scales = 'free_x', space = 'free_x') + 
  geom_point() + ggtitle("PI ratio of RADseq SNPs to their neghbourhood") +
  xlab("Position [Mb]") + ylab("PI ratio to neighbourhood") + guides(col=F)
```

```{r RAD_PI2, cache=T, cache.lazy=F, dependson='RAD_PI', echo=F, fig.cap="\\label{fig:neigh_PI}PI values of the local (10kb) neighbourhood of each RADseq SNP. The color of each SNP shows whether the -log10 of the BH-corrected p-value from the association mapping for CSD was below 10e-5."}
ggplot(data=snps, aes(x=BP/zoom, y=around10k, col=fisher<5)) + 
  facet_grid(~Chr, scales = 'free_x', space = 'free_x') + 
  geom_point() + ggtitle("PI of RADseq SNPs neghbourhood")+
  xlab("Position [Mb]") + ylab("PI") + guides(col=F)
```


```{r up_down_PI10k, dependson='RAD_PI', cache=T, cache.lazy=F, echo=F, fig.cap="\\label{fig:up_down}PI values of the local (10kb) neighbourhood of each significant RADseq SNP compared to upstream and downstream regions. Regions belonging to a SNP are linked by a segment."}

# Comparing PI value around RADseq candidates vs upstream and downstream background
snps$up = snps$down <- NA
for(chr in unique(RAD$Chr)){
  PI_chr <- PI_sites %>% filter(Chr==chr)
  snps$up[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x+around) & 
                                                                     PI_chr$BP < (x+2*around)], na.rm=T))
  snps$down[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x-2*around) & 
                                                                     PI_chr$BP < (x-around)], na.rm=T))
}
sig <- snps[snps$fisher >=5,]
long_sig <- gather(sig, position, PI, around10k:up, factor_key=TRUE)
long_sig$position <- factor(long_sig$position, ordered = T, levels = c("up", "around10k", "down"))

ggplot(data=long_sig, aes(x=position, y=PI)) +
  geom_boxplot() + 
  geom_segment(data=sig, aes(x=1,y=up,xend=2,yend=around10k)) + 
  geom_segment(data=sig, aes(x=2,y=around10k,xend=3,yend=down)) +
  geom_point(aes(col=Chr))+
  scale_x_discrete(labels=c("Upstream\n[-20k:-10k]", "Candidate\n[-10k:+10k]", "Downstream\n[+10k:+20k]")) +
  theme_bw()

```

```{r up_down_PI1k, dependson='RAD_PI1k', cache=T, cache.lazy=F, echo=F, fig.cap="\\label{fig:up_down}PI values of the local (10kb) neighbourhood of each significant RADseq SNP compared to upstream and downstream regions. Regions belonging to a SNP are linked by a segment."}
around <- 1000
# Comparing PI value around RADseq candidates vs upstream and downstream background
snps$up = snps$down <- NA
for(chr in unique(RAD$Chr)){
  PI_chr <- PI_sites %>% filter(Chr==chr)
  snps$up[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x+around) & 
                                                                     PI_chr$BP < (x+2*around)], na.rm=T))
  snps$down[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x-2*around) & 
                                                                     PI_chr$BP < (x-around)], na.rm=T))
}
sig <- snps[snps$fisher >=5,]
long_sig <- gather(sig, position, PI, around10k:up, factor_key=TRUE)
long_sig$position <- factor(long_sig$position, ordered = T, levels = c("up", "around10k", "down"))

ggplot(data=long_sig, aes(x=position, y=PI)) +
  geom_boxplot() + 
  geom_segment(data=sig, aes(x=1,y=up,xend=2,yend=around10k)) + 
  geom_segment(data=sig, aes(x=2,y=around10k,xend=3,yend=down)) +
  geom_point(aes(col=Chr))+
  scale_x_discrete(labels=c("Upstream\n[-2k:-1k]", "Candidate\n[-1k:+1k]", "Downstream\n[+1k:+2k]")) +
  theme_bw()

```

```{r up_down_PI100, dependson='RAD_PI100', cache=T, cache.lazy=F, echo=F, fig.cap="\\label{fig:up_down}PI values of the local (10kb) neighbourhood of each significant RADseq SNP compared to upstream and downstream regions. Regions belonging to a SNP are linked by a segment."}
around <- 100
# Comparing PI value around RADseq candidates vs upstream and downstream background
snps$up = snps$down <- NA
for(chr in unique(RAD$Chr)){
  PI_chr <- PI_sites %>% filter(Chr==chr)
  snps$up[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x+around) & 
                                                                     PI_chr$BP < (x+2*around)], na.rm=T))
  snps$down[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(PI_chr$PI[PI_chr$BP > (x-2*around) & 
                                                                     PI_chr$BP < (x-around)], na.rm=T))
}
sig <- snps[snps$fisher >=5,]
long_sig <- gather(sig, position, PI, around10k:up, factor_key=TRUE)
long_sig$position <- factor(long_sig$position, ordered = T, levels = c("up", "around10k", "down"))

ggplot(data=long_sig, aes(x=position, y=PI)) +
  geom_boxplot() + 
  geom_segment(data=sig, aes(x=1,y=up,xend=2,yend=around10k)) + 
  geom_segment(data=sig, aes(x=2,y=around10k,xend=3,yend=down)) +
  geom_point(aes(col=Chr))+
  scale_x_discrete(labels=c("Upstream\n[-200:-100]", "Candidate\n[-100:+100]", "Downstream\n[+100:+200]")) +
  theme_bw()

```

After discussion, I added the following figure showing the PI values around SNPs versus up- and downstream regions, but stopping windows after they reach N SNPs rather than just basepair distance

```{r up_down_nSNP, dependson='RAD_PI100', cache=T, cache.lazy=F, echo=F, fig.cap="\\label{fig:up_down_snp}PI values of the local (10 closest SNPs) neighbourhood of each significant RADseq SNP compared to windows of 10 SNPs 10kb upstream and downstream. Regions belonging to a candidate are linked by a segment."}
around <- 10000
N_snp <- 10

snp_window <- function(bp, N){
  # Subsets the PI values chromosome to a window centered around bp containing N SNPs.
  win_snp <- snp_BP[order(abs(bp-snp_BP))][1:N]
  win_PI <- PI_chr$PI[PI_chr$BP < max(win_snp) & PI_chr$BP > min(win_snp)]
  return(win_PI)
}

# Comparing PI value around RADseq candidates vs upstream and downstream background
snps$up = snps$down <- NA
for(chr in unique(RAD$Chr)){
  
  # Will be used by snp_window
  snp_idx <- which(PI_chr$PI>0)
  snp_BP <- PI_chr$BP[snp_idx]
  
  PI_chr <- PI_sites %>% filter(Chr==chr)
  snps$up[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(snp_window(x + around, N_snp), na.rm=T))
  snps$down[snps$Chr==chr] <- sapply(snps$BP[snps$Chr==chr], 
                                          function(x) mean(snp_window(x - around, N_snp), na.rm=T))
}

sig <- snps[snps$fisher >=5,]
long_sig <- gather(sig, position, PI, around10k:up, factor_key=TRUE)
long_sig$position <- factor(long_sig$position, ordered = T, levels = c("up", "around10k", "down"))

ggplot(data=long_sig, aes(x=position, y=PI)) +
  geom_boxplot() + 
  geom_segment(data=sig, aes(x=1,y=up,xend=2,yend=around10k)) + 
  geom_segment(data=sig, aes(x=2,y=around10k,xend=3,yend=down)) +
  geom_point(aes(col=Chr))+
  facet_wrap(~Chr)+
  scale_x_discrete(labels=sprintf(c("Upstream\n[%i SNPs,-%ikb]", "Hit\n[%i SNPs, Hit]", "Downstream\n[%i SNPs, +%ikb]"),N_snp,around/1000)) +
  theme_bw()

```