---
title: "Nucleotidic diversity in wild samples"
author: "Cyril Matthey-Doret"
date: "19/02/2018"
output: pdf_document
---

```{r setup, echo=F, include=F}
packs <- c("ggplot2","Rwave","dplyr", "gridExtra")
lapply(packs, require, character.only=T)
```

# Data
Paired end whole genome sequencing data from 11 males 4 females. Each sample has between 14 and 25 million pairs.

```{bash karyotype, echo=F}

# Calculating size of all chromosomes and storing them in a file
cd "/Users/cyril/Documents/UNIL/Master/master_project/"
ref_path="data/ref_genome/ordered_genome/merged.fasta"
cat $ref_path | awk '$0 ~ ">" {print c; c=0;printf substr($0,2,100) "\t"; } $0 !~ ">" {c+=length($0);} END { print c; }' | grep "chr" > "data/ref_genome/ordered_genome/karyotype.txt"
```


```{r loading, echo=F}

setwd("~/Documents/UNIL/Master/master_project/")
# Coordinates of CSD GWAS hits
csd_hits <- read.table("data/assoc_mapping/case_control/case_control_hits.tsv", header = T, stringsAsFactors = F)

# Anchored contigs containing CSD hits
csd_contig <- read.table("data/assoc_mapping/CSD_tig.tsv", stringsAsFactors = F)
# Renaming headers to match those of csd_hits
colnames(csd_contig) <- c("CHROM", "START", "END")

# Nucleotidic diversity windows along wild mothers genomes
pi <- read.table("data/wgs_wild_mothers/stats/nucleo_div.windowed.pi", sep="\t", header=TRUE, stringsAsFactors = F)

# List of chromosomes with their respective lengths
karyo <- read.table("data/ref_genome/ordered_genome/karyotype.txt", stringsAsFactors = F)
  
```
# Status

## Done

* Aligned PE reads to the genome (BWA-mem).
* SNP calling (bcftools)
    + Split genome in 100kb regions.
    + Perform SNP calling with all 15 samples on each region separately (mpileup).
    + Keep only variant sites (bcftools call)
    + Merge VCF files into a single file for the whole genome.
* Compute nucleotidic diversity PI in 100bp windows.
* Compare PI distributions
    + csd vs non-csd regions (Figure 1).
    + among different csd regions (Figure 2).

## In progress

* Use only max (or nth top percentile ?) windows to identify relevant CSD windows.

## To do

* Heterozygosity (check male vs females).
* Mapping statistics.
* Tajima's D to compare distribution of PI in CSD regions vs neutral expectations.
* Remove PCR duplicate reads (Picard).
* Filter variants to reduce noise: (vcftools).
    + Repetitive regions: max coverage filter.
    + Wrong calls: low coverage filter.
    + Low information sites: called in at least X% of samples.

# Optional

* Use wavelet transform or similar method to select optimal PI window size.

```{r filtering, echo=F}

# Filtering out unanchored contigs
pi <- pi[grep(pattern="chr", pi$CHROM),]

# Filtering significant (p<10-3) hits. fisher column represents -log10 pval
csd_hits <- csd_hits %>%
  rename(CHROM=Chr) %>%
  filter(fisher >= 3)
```

```{r fig1:summary_pi, echo=F}

# CSD windows around significant SNPs
win_size <- 100000
csd_win <- csd_hits %>%
  select(CHROM, BP) %>%
  mutate(START=BP-win_size/2, END=BP+win_size/2, ID=row_number()) %>%
  select(-BP)

in_win <- function(rnum, dfWin, dfReg){
  # fetches PI windows overlapping with CSD region at row number rnum.
  tmpdf <- dfWin[dfWin$CHROM == dfReg[rnum, "CHROM"] & 
         dfWin$BIN_START > dfReg[rnum, "START"] &
         dfWin$BIN_END < dfReg[rnum, "END"],]
  if(nrow(tmpdf)){
    return(cbind(tmpdf,ID=dfReg$ID[rnum]))
  }
}

# Finding all PI windows overlapping CSD regions
tmp <- sapply(1:nrow(csd_win), function(x) in_win(x, dfWin=pi, dfReg=csd_win), simplify = F)
csd_pi <- do.call(rbind,tmp)
csd_pi <- csd_pi[!duplicated(csd_pi),]

# Show Mb instead of basepairs
zoom=1000000

# Plotting all PI windows along genome in blue. Windows in CSD regions are in red.
p0 <-ggplot(pi, aes(x=BIN_START/zoom, y=PI)) + 
  facet_grid(CHROM~., 
             space = 'free_x', 
             scales='free_x') + 
  geom_rect(data=csd_win, 
            inherit.aes = F, 
            aes(xmin=START/zoom, xmax=END/zoom, ymin=0, ymax=max(pi$PI)), alpha=0.3) +
  geom_point(col="blue", size=0.5) + geom_point(data=csd_pi, col="red", size=0.6) +
  xlab("Mb")
  
# Histogram of PI values in windows outside CSD regions (blue) versus inside (red)
p1 <- ggplot(data=pi, aes(x=PI)) + 
  geom_histogram(fill="blue", aes(x=PI, ..density..), alpha=0.5, binwidth = 0.001) + 
  geom_histogram(data=csd_pi, fill="red", aes(x=PI, ..density..), alpha=0.3, binwidth = 0.001) +
  theme_bw()

# Same with boxplots
p2 <- ggplot(data=pi, aes(x=1, y=PI)) + 
  geom_boxplot(fill="blue") + 
  geom_boxplot(data=csd_pi, fill="red", aes(x=2)) + 
  coord_flip() + theme_minimal() + xlab("")

grid.arrange(grobs=list(p0, p1,p2), layout_matrix=rbind(c(1,1,2),
                                                        c(1,1,2),
                                                        c(1,1,3)))

```

```{r fig2:compare_pi, echo=F}

# Comparing distributions of PI windows in whole genome and each CSD region independently
pi$ID <- 0
csd_pi$ID <- as.factor(csd_pi$ID)

ggplot(data=pi, aes(group=ID, x=as.numeric(ID), y=PI)) +
  geom_boxplot(fill="blue") + 
  geom_boxplot(data=csd_pi, fill="red") + 
  coord_flip() + theme_minimal()

```

```{r fig3:subset_pi, echo=F}

# Compute 95th percentile and maximum PI in each CSD region to rule out those with only low values
highPI <- csd_pi %>% 
  group_by(ID) %>%
  summarise(p95=quantile(PI, probs=0.95), max=max(PI))

# Some p95 are higher, but is this difference relevant ? Max likely irrelevant.
ggplot() + 
  geom_point(data=highPI, aes(x=p95, y=max), col="red") +
  geom_label(data=highPI, aes(x=p95, y=max, label=ID), col="red") +
  geom_label(aes(x=quantile(pi$PI, probs=0.95), y=max(pi$PI), label="Genome"), col="blue") +
  xlab("95th percentile of PI") + ylab("Maximum PI") + theme_bw()

```

```{r bootstrap, echo=F}

# Random sampling of equal size regions and compute p95 in each -> generate empirical distribution

randRegion <- function(regSize, chroms=karyo) {
  # Generate random region
  # Select chromosome (random choice weighted by their length)
  selChr <- sample(x = chroms[,1],
                   prob = chroms[,2], 
                   size = 1, 
                   replace = T)
  # Total length of selected chromosome
  selLen <- chroms[chroms[,1]==selChr,2]
  # Choose random position for start of region in the chromosome (need to fit entire region)
  selPos <- round(runif(n = 1,
                        min = 1,
                        max = (selLen-regSize)))
  
  return(c(selChr, selPos))
}

# Generate many regions
regGen <- data.frame(t(replicate(randRegion(win_size), n = 10000)), stringsAsFactors = F)
regGen[,2] <- as.numeric(regGen[,2])
regGen[,3] <- regGen[,2] + win_size
regGen[,4] <- 1:nrow(regGen)
colnames(regGen) <- c("CHROM","START","END", "ID")

# List all PI windows contained in bootstrap regions
pi <- pi[,-6]
tmp <- sapply(1:nrow(regGen), function(x) in_win(x, dfWin = pi, dfReg = regGen), simplify = F)
boot_pi <- do.call(rbind,tmp)
boot_highPI <- boot_pi %>% 
  group_by(ID) %>%
  summarise(p95=quantile(PI, probs=0.95), max=max(PI))

# Some p95 are higher, but is this difference relevant ? Max likely irrelevant.
ggplot() + 
  geom_histogram(data=boot_highPI, aes(x=p95), fill="blue") +
  geom_vline(data=highPI, aes(xintercept=p95),col="red") +
  xlab("95th percentile of PI") + theme_bw()

```

```{r fig4:wavelet, echo=F}
#pi <- pi[pi$CHROM=="chr3",]
#pi_cwt <- cwt(input = pi$PI, noctave = 4, nvoice = 8, plot = T)
#plot(pi_cwt[,1])
```