---
title: "computing PI"
author: "Cyril Matthey-Doret"
date: "28/03/2018"
output: pdf_document
---

# Basic method

## Pseudocode implementation

Reused vcftools algorithm:
* Bases that are not called (NA) will not be counted in pairs
* Fixed bases that are called will count in pairs and decrease pi

```{r eval=F}
pairs = 0
mismatch = 0

for pos in sequence:
  for allele in pos:
    n_other = pos.freq[all] - pos.freq[allele]
    mismatch += pos.freq[allele] * n_other
  pairs += pos.freq[all] * (pos.freq[all] - 1)
  pi[pos] = mismatch/pairs
```

## Variants

* Computing PI per site or in windows ($\frac{sum(mismatch)}{sum(pairs)}$ in each window)
* Including only SNPs vs all fixed sites and SNPs.
* Weighting windows per total number of alleles (only useful if using a smoother)

# Results

## Note on non-anchored contigs
Checked for GWAS hits on non-anchored contigs again after adding MAF filter.

## Different window sizes

##