---
title: 'New sequencing data: Early results'
author: "Cyril Matthey-Doret"
date: "08/10/2017"
header-includes:
   - \usepackage{placeins}
output: pdf_document
---

# General statistics

## Number of individuals
From the 768 processed samples, 704 made it to the end of the pipeline. Out of those, there were 365 haploid males. Among the 339 diploid individuals left, there are 176 males and 163 females. There is a total of 60 families among which 48 contain diploid individuals and 37 contain both diploid males and females.

## Homozygosity values
Proportion of homozygosity among variant sites was used to classify individuals as haploid or diploid. The figure below represents the distribution of homozygosity. across all individuals

```{r hom_dist, echo=FALSE, fig.cap="Distribution of the proportion of homozygosity among variant sites per individuals.", out.width='60%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/hom_dist.pdf')
```

# Grouping families

The goal in this section is to split families into groups that reflect the CSD genotype of the mothers. I attempted to achieve this by clusterint families according to the proportion of males among diploid offspring. Because we only sequence a fraction of the offspring in each family, I computed the rate or haploidy among sequenced males in each family and extrapolated this proportion to all (non-sequenced) males to infer the total number of diploid males. I then used these values to measure the ratio of diploid males to diploid offspring.

The expected number of groups depend on the number of CSD loci we consider. Each group will reflect the number of heterozygous CSD loci in the mothers as follows.

```{r prop_gr, echo=FALSE, fig.cap="Possible mother genotypes of mothers with associated proportion of diploid males when considering different number of CSD loci. The proportion of diploid males produced by a mother (Pd) depends on the recombination rate $r_x$ at the heterozygous CSD loci.", out.width='60%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/prop_table.pdf')
```

I clustered families automatically using the k-means methods in order to obtain unbiased groups. I considered 2 different scenarios: 2 CSD loci on different chromosomal arms and 3 CSD loci on different chromosomal arms. I don't have enough families to detect scenarios with more loci. The clusters are not clearly separated and this could be due to: inaccurate metric, high number of loci, high variability in recombination rates... ?


```{r groups, echo=FALSE, fig.cap="Family groups obtained through k-means clustering based on inferred proportion of males among diploid offspring. top: scenario considering 2 loci on different chromosomal arms, genotypes corresponding to the clusters are displayed in the legend. bottom: scenario considering 3 loci on different chromosomal arms. Genorypes are not displayed as they will depend on the relative recombination rates of the loci. Piecharts show the number of families belonging to each cluster.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/full_groups.pdf')
```

\FloatBarrier

# Association mapping

I performed association mapping both pooling all families together, or splitting them according to the two scenario above. I use one-sided Fisher exact test on each locus to test if there is a there is a positive association between heterozygosity at the locus and female sex.

## 2 loci

```{r C2L_nocorr, echo=FALSE, fig.cap="Association mapping considering 2 loci scenario. No correction for multiple testing was applied.", out.width='80%',fig.pos='h!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_2loci_nocorr.pdf')
```


```{r C2L_BH, echo=FALSE, fig.cap="Association mapping considering 2 loci scenario. Benjamini-Hochberg correction for multiple testing was applied.", out.width='80%',fig.pos='h!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_2loci_BH.pdf')
```

\FloatBarrier

## 3 loci
Note that group 7 is always absent because it only contains haploid mals and daughters. -> sexual mothers ?

```{r C3L_nocorr, echo=FALSE, fig.cap="Association mapping considering 3 loci scenario. No correction for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_3loci_nocorr.pdf')
```


```{r C3L_BH, echo=FALSE, fig.cap="Association mapping considering 3 loci scenario. Benjamini-Hochberg correction for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_3loci_BH.pdf')
```

\FloatBarrier

## Pooling all families


```{r CP_nocorr, echo=FALSE, fig.cap="Association mapping pooling all families together. No correction for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_pooled_nocorr.pdf')
```


```{r CP_BH, echo=FALSE, fig.cap="Association mapping pooling all families together. Benjamini-Hochberg for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_pooled_BH.pdf')
```

```{r CP_bonferroni, echo=FALSE, fig.cap="Association mapping pooling all families together. Bonferroni for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_pooled_bonferroni.pdf')
```

### Non-mapped contigs

```{r CPU_BH, echo=FALSE, fig.cap="Association mapping pooling all families together showing unordered contigs only. Benjamini-Hochberg correction for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_pool_unord_BH.pdf')
```

```{r CPU_bonferroni, echo=FALSE, fig.cap="Association mapping pooling all families together showing unordered contigs only. Bonferroni for multiple testing was applied.", out.width='80%',fig.pos='htb!', fig.align='center'}
knitr::include_graphics('./figures/association_mapping/case_pool_unord_bonferroni.pdf')
```

\FloatBarrier


# Ideas

More candidates than expected when pooling families, none when clustering. The clustering is probably wrong and different genotypes are mixed in the same groups.

What I could do: Blast candidate regions. Look at what kind of genes are in there. Check for homologies across regions and with other hymenoptera.

Possible to improve genome using my samples ? Linkage map etc (ALLMAPS)

Fix clustering problem using different metric ? (but what, then ?)


# Pipeline parameters

mapping: BWA aln, Mismatches=4

```{r mapstats, echo=F}
mapval <- c(170435458,94668507,44531856,31235095)
maptab <- rbind(c("tot","single","multi","miss"), 
                c(mapval[1],
                  round(mapval[2]/mapval[1], 3),
                  round(mapval[3]/mapval[1], 3),
                  round(mapval[4]/mapval[1], 3)))
knitr::kable(maptab)
```

pstacks: mininum stack coverage=3

Stats averaged across samples.

```{r Pstats, echo=F}
Psval <- c(2694.0,50.0,69.4)
Pstab <- rbind(c("N loci","mean coverage","SD coverage"), 
                c(Psval[1],
                  round(Psval[2], 3),
                  round(Psval[3], 3)))
knitr::kable(Pstab)
```



cstacks: Mismatch=3

```{r Cstats, echo=F}
Csval <- c(12585, 56971)
Cstab <- rbind(c("N Loci", "N alleles"), 
                c(Csval[1],Csval[2]))
knitr::kable(Cstab)
```


populations: Min proportion (R)=0.8, min depth=5

```{r pop_stats, echo=F, message=F}
library(tidyverse)
pop_tbl <- read.table("../../../data/ploidy/thresholds/fixed_grouped_d-5_r-80.tsv", header=T)
sum_pop <- pop_tbl %>% 
  summarise(avg_depth=mean(MEAN_DEPTH), med_depth=median(MEAN_DEPTH), sd_depth=sd(MEAN_DEPTH), 
            avg_SITES=mean(N_SITES), med_sites=median(N_SITES), sd_sites=sd(N_SITES))

colnames(sum_pop) <- c("mean depth", "median depth", "sd depth", "mean N sites", "median N sites", "sd N sites")
knitr::kable(sum_pop)

```
