---
title: "Alleles segregation"
author: "Cyril Matthey-Doret"
date: "09/04/2018"
output: pdf_document
---

# Intro

To look into the segregation of alleles at peaks of allelic diversity. I will be comparing the results of PCA at the whole genome level, versus in the individual regions. 


```{r setup, echo=F}
packs <- c("ggplot2","dplyr","readr","viridis", "magrittr", "gridExtra", "ggrepel", "FactoMineR", "factoextra")
packs <- suppressPackageStartupMessages(sapply(packs, library, quietly=T, character.only=T))

setwd("../../../../")
raw_snp <- read_tsv("data/wgs_wild/variant/snp.chr.wild.matrix.txt", col_names=F, na='.', col_types = cols())
# Remove SNPs where non-reference allele is fixed
snp <- raw_snp %>% filter_at(-(1:2), any_vars(.!=1))
# Extract SNPs with more than 2 alleles as "alt_snp"
alt_snp <- snp %>% filter_at(-(1:2), any_vars(.>1))

# Sample names for plotting
sample_names <- read.table("data/wgs_wild/wgs_samples.tsv", header=F, stringsAsFactors = F)
colnames(sample_names) <- c("Name","Sex")
# Generating separate ID for second haplotype of females
tmp_fem <- sample_names %>% filter(Sex=="F") %>% mutate(Name = sprintf("%s_2", Name))
sample_names <- sample_names %>% 
  mutate(Name = ifelse(Sex=='F', 
                       yes=sprintf("%s_1",Name),
                       no=Name)) %>%
  bind_rows(., tmp_fem) %>%
  arrange(Name)

zoom <- 10^6
```


# Preprocessing

Fixed sites are not useful to characterize genetic distances between individuals. To only include sites that are variable within the current sample, I removed all ```r raw_snp %>% filter_at(-(1:2), all_vars(.==1)) %>% nrow(.)``` sites where an alternative allele was fixed, leaving a total of ```r nrow(snp)```. When performing PCA on a genotype matrix there are two possibilities to encode alleles. 

```{r N_SNPs, echo=F, warning=F}
alleles <- apply(snp[,-c(1,2)], MARGIN = 1, FUN=max, na.rm=T)
# Alleles are 0-indexed: number of alleles = max(alleleID) + 1
alleles <- data.frame(cbind(Chr=snp$X1, N_alleles=alleles+1))

ggplot(alleles, aes(x=N_alleles)) + 
  geom_histogram(stat="count") +
  facet_wrap(~Chr) + 
  scale_y_log10() + 
  ggtitle("Distribution of SNPs per number of alleles")
```

1. All non-reference alleles can be pooled together and the metric used is the number of non-reference alleles.
2. All alternative alleles are kept as separate "dummy" variables and the metric is the presence/absence of each allele.

Because the samples we have show peaks of nucleotidic diversity and are quite diverse, I will first check if the regions with more than one alternative allele seem to be concentrated at some points or if they follow about the same distribution as all other SNPs. If there is no hotspot, I will consider pooling all alternative alleles to simplify the PCA analysis.

```{r distribution, echo=F, fig.height = 9, fig.width = 6}
alt_size <- alt_snp %>% 
  rename(Chr=X1) %>% 
  group_by(Chr) %>% 
  summarise(N=n())

dist_all <- ggplot(data=snp, aes(x = X2/zoom)) + 
  geom_histogram(binwidth = 100000/zoom) + 
  facet_grid(~X1) + 
  theme_bw() + 
  xlab("") +
  ylab("All SNPs")

dist_alt <- ggplot(data=alt_snp, aes(x = X2/zoom)) + 
  geom_histogram(binwidth = 100000/zoom) + 
  facet_grid(~X1) + 
  theme_bw() + 
  xlab("Genomic position [Mb]") + 
  ylab("SNPs with >2 alleles")

comp_dist <- snp %>%
  group_by(X1) %>%
  arrange(X2) %>%
  do(
    data.frame(approx(x = 1L:(nrow(.)), 
                      y = .$X2, 
                      n = alt_size[[which(alt_size$Chr==.$X1[1]), "N"]])$y)
    ) %>% 
  ungroup(X1) %>% 
  mutate(j = alt_snp$X2)

colnames(comp_dist) <- c("Chr","sim_SNP_BP","alt_SNP_BP")

qq_comp <- ggplot(data=comp_dist, aes(x=sim_SNP_BP/zoom, y=alt_SNP_BP/zoom)) + 
  geom_line() + 
  facet_wrap(~Chr) +
  geom_abline(slope=1,intercept=0, col='grey60', lty=2) + 
  theme_bw() + 
  ggtitle("Repartition of diverse (>2 alleles) SNPs throughout the genome") + 
  xlab("Interpolated repartition of all SNPs [Mb]") + 
  ylab("Observed repartition of SNPs with >2 alleles [Mb]")

grid.arrange(layout_matrix=rbind(c(1,1),
                                 c(2,2),
                                 c(3,3),
                                 c(3,3)),
             grobs=list(dist_all, dist_alt, qq_comp))
```

```{r Hamming_MDS, echo=F}



hamming <- function(X, Y){
    if (missing(Y)){
        uniqs <- unique(as.vector(X))
        U <- X == uniqs[1]
        H <- t(U) %*% U
        for(uniq in uniqs[-1]){
            U <- X == uniq
            H <- H + t(U) %*% U
        }
    } else{
        uniqs <- union(X, Y)
        H <- t(X == uniqs[1]) %*% (Y == uniqs[1])
        for (uniq in uniqs[-1]){
            H <- H + t(X == uniq) %*% (Y == uniq)
        }
    }
    nrow(X) - H
}

mds_snp <- as.matrix(snp[,-c(1,2)])
mds_snp <- mds_snp[complete.cases(mds_snp),]

H <- hamming(mds_snp)
H_MDS <- cmdscale(H)
colnames(H_MDS) <- c("Dim1","Dim2"); rownames(H_MDS) <- NULL
H_MDS <- cbind(H_MDS, sample_names)

ggplot(H_MDS, aes(x=Dim1, y=Dim2, col=Sex, label=Name)) + 
  geom_point() + 
  theme_bw() + 
  geom_text_repel(size=2.5) + 
  ggtitle("MDS on Hamming distance matrix of samples using all SNPs")

```

```{r pca, echo=F}

res.pca <- PCA(t(mds_snp), graph = FALSE, scale.unit = F)
ind_pca <- cbind(sample_names, res.pca$coord)
rownames(ind_pca) <- NULL

# Percentage of variance explained by each PC
#scree <- fviz_screeplot(res.pca, ncp=10)
# Samples on 2 first PCs
ggplot(ind_pca, aes(x=Dim.1, y=Dim.2, col=Sex, label=Name)) + 
  geom_point() + 
  theme_bw() + 
  geom_text_repel(size=2.5) + 
  ggtitle("PCA on all SNPs") + 
  xlab("PC1") + 
  ylab("PC2") +


```