---
title: "Population structure"
author: "Cyril Matthey-Doret"
date: "09/04/2018"
output: pdf_document
---

# NOTE: Female haplotypes are separated to compute distances in this report. This is wrong and will be addressed in the next one

# Intro

To look into the segregation of alleles at peaks of allelic diversity. I will be comparing the results of PCA at the whole genome level, versus in the individual highly diverse regions. 


```{r setup, echo=F}
packs <- c("ggplot2","dplyr","readr","viridis", "magrittr", "gridExtra", "ggrepel", "FactoMineR", "factoextra")
packs <- suppressPackageStartupMessages(sapply(packs, library, quietly=T, character.only=T))

setwd("../../../../")
raw_snp <- read_tsv("data/wgs_wild/variant/snp.chr.wild.matrix.txt", col_names=F, na='.', col_types = cols())
# Remove SNPs where non-reference allele is fixed
snp <- raw_snp %>% filter_at(-(1:2), any_vars(.!=1))
# Extract SNPs with more than 2 alleles as "alt_snp"
alt_snp <- snp %>% filter_at(-(1:2), any_vars(.>1))

# Sample names for plotting
sample_names <- read.table("data/wgs_wild/wgs_samples.tsv", header=F, stringsAsFactors = F)
colnames(sample_names) <- c("Name","Sex")
# Generating separate ID for second haplotype of females
tmp_fem <- sample_names %>% filter(Sex=="F") %>% mutate(Name = sprintf("%s_2", Name))
sample_names <- sample_names %>% 
  mutate(Name = ifelse(Sex=='F', 
                       yes=sprintf("%s_1",Name),
                       no=Name)) %>%
  bind_rows(., tmp_fem) %>%
  arrange(Name)

zoom <- 10^6

# Nucleotidic diversity (PI) computed in 100bp windows
div100 <- read_tsv("data/wgs_wild/stats/win_w100_t10_PI.tsv", col_type=cols(), col_names=F)
```


# Preprocessing

Fixed sites are not useful to characterize genetic distances between individuals. To only include sites that are variable within the current sample, I removed all ```r raw_snp %>% filter_at(-(1:2), all_vars(.==1)) %>% nrow(.)``` sites where an alternative allele was fixed, leaving a total of ```r nrow(snp)```. I will try to strategies to visualize genetic distances between individuals.

Note that for all statistical analysis (PCA and MDS), I removed all SNPs where any individual had a missing call. This should not affect the analysis too much as there are only ```r raw_snp %>% summarise```

# Whole genome analyis
Trying two separate methods for visualizing relatedness between samples.

## MDS on Hamming distance matrix
MDS is equivalent to PCA but preserves distances rather than variance. It can be performed on a distance matrix. Hamming distance is a metric that counts the number of substitution between two strings of equal length. An all vs all SNP distance matrix using Hamming distance allows to perform MDS and keep the information of alternative alleles.

## PCA with pooled alleles
Because the number of SNPs with more than two alleles is low (Figure \ref{fig:N_SNPs}), I should be able to pool all alternative alleles (i.e. encode all alleles `1`,`2` and `3` as `1`) without affecting the results too much. SNPs with more than 2 seem to follow approximately the same distribution as all SNPs and not be concentrated in "hotspots" *(Figure \ref{fig:dist_SNPs}, Discuss QQplots with Tanja)*.

PCA yields similar results to MDS on hamming distance matrix and the implementation is much slower, then I will stick with Hamming distance + MDS for visualization (Figures \ref{fig:MDS} and \ref{fig:PCA}). Both methods show a high divergence of haplotypes in females and a strong clustering of males. Could that somehow an artifact / mistake ?

```{r N_SNPs, cache=T, dependson="setup", echo=F, warning=F, fig.cap="\\label{fig:N_SNPs}Number of SNPs with 2, 3 and 4 alleles on each chromosome. Note the logarithmic scale. There are no SNPs with more than 4 alleles (or at least they were not called.)"}

# Alleles are 0-indexed: number of alleles = max(alleleID) + 1 except if reference allele (0) is not present.
# subtracting min(alleleID) accounts for that.
alleles <- apply(snp[,-c(1,2)], MARGIN = 1, function(x) (max(x, na.rm=T) - min(x, na.rm=T)) + 1)
alleles <- data.frame(cbind(Chr=snp$X1, N_alleles=alleles))

ggplot(alleles, aes(x=N_alleles)) + 
  geom_histogram(stat="count") +
  facet_wrap(~Chr) + 
  scale_y_log10() + 
  ggtitle("Distribution of SNPs per number of alleles") + 
  xlab("Number of alleles")

```


```{r distribution, cache=T, dependson="setup", echo=F, fig.height = 9, fig.width = 6, fig.cap="\\label{fig:dist_SNPs}Genomic repartition of all SNPs and SNPs with more than 2 alleles (top) and direct comparison of the distributions using QQplots (bottom)."}
alt_size <- alt_snp %>% 
  rename(Chr=X1) %>% 
  group_by(Chr) %>% 
  summarise(N=n())

dist_all <- ggplot(data=snp, aes(x = X2/zoom)) + 
  geom_histogram(binwidth = 100000/zoom) + 
  facet_grid(~X1) + 
  theme_bw() + 
  xlab("") +
  ylab("All SNPs")

dist_alt <- ggplot(data=alt_snp, aes(x = X2/zoom)) + 
  geom_histogram(binwidth = 100000/zoom) + 
  facet_grid(~X1) + 
  theme_bw() + 
  xlab("Genomic position [Mb]") + 
  ylab("SNPs with >2 alleles")

comp_dist <- snp %>%
  group_by(X1) %>%
  arrange(X2) %>%
  do(
    data.frame(approx(x = 1L:(nrow(.)), 
                      y = .$X2, 
                      n = alt_size[[which(alt_size$Chr==.$X1[1]), "N"]])$y)
    ) %>% 
  ungroup(X1) %>% 
  mutate(j = alt_snp$X2)

colnames(comp_dist) <- c("Chr","sim_SNP_BP","alt_SNP_BP")

qq_comp <- ggplot(data=comp_dist, aes(x=sim_SNP_BP/zoom, y=alt_SNP_BP/zoom)) + 
  geom_line() + 
  facet_wrap(~Chr) +
  geom_abline(slope=1,intercept=0, col='grey60', lty=2) + 
  theme_bw() + 
  ggtitle("Repartition of diverse (>2 alleles) SNPs throughout the genome") + 
  xlab("Interpolated repartition of all SNPs [Mb]") + 
  ylab("Observed repartition of SNPs with >2 alleles [Mb]")

grid.arrange(layout_matrix=rbind(c(1,1),
                                 c(2,2),
                                 c(3,3),
                                 c(3,3)),
             grobs=list(dist_all, dist_alt, qq_comp))
```

```{r Hamming_MDS, cache=T, dependson="setup", echo=F, fig.cap="\\label{fig:MDS}MDS on hamming distance matrix using all SNPs with any number of alleles. Samples are colored according to sex and IDs are shown directly on the plot. For diploid samples (females), haplotypes are encoded as _1 and _2."}



hamming <- function(X, Y){
  # Fast implementation of Hamming distances using matrix dot products.
  # Written by Johann de Jong
    if (missing(Y)){
        uniqs <- unique(as.vector(X))
        U <- X == uniqs[1]
        H <- t(U) %*% U
        for(uniq in uniqs[-1]){
            U <- X == uniq
            H <- H + t(U) %*% U
        }
    } else{
        uniqs <- union(X, Y)
        H <- t(X == uniqs[1]) %*% (Y == uniqs[1])
        for (uniq in uniqs[-1]){
            H <- H + t(X == uniq) %*% (Y == uniq)
        }
    }
    nrow(X) - H
}

mds_snp <- as.matrix(snp[,-c(1,2)])
mds_snp <- mds_snp[complete.cases(mds_snp),]

H <- hamming(mds_snp)
H_MDS <- cmdscale(H)
colnames(H_MDS) <- c("Dim1","Dim2"); rownames(H_MDS) <- NULL
H_MDS <- cbind(H_MDS, sample_names)

ggplot(H_MDS, aes(x=Dim1, y=Dim2, col=Sex, label=Name)) + 
  geom_point() + 
  theme_bw() + 
  geom_text_repel(size=2.5) + 
  ggtitle("MDS on Hamming distance matrix of samples using all SNPs")

```

```{r pca, cache=T, dependson="Hamming_MDS", echo=F, fig.cap="\\label{fig:PCA}PCA all SNPs with all alternative alleles pooled. Samples are colored according to sex and IDs are shown directly on the plot. For diploid samples (females), haplotypes are encoded as _1 and _2"}
# note NAs have been removed already in MDS chunk
pca_snp <- mds_snp
# Ensuring all SNPs have alleles 0 and 1 (and not 1/2 in case ref allele is absent)
pca_snp <- matrix(apply(pca_snp, MARGIN = 1, function(x) x-min(x)), ncol=ncol(pca_snp), byrow = TRUE)
pca_snp[pca_snp > 1] <- 1

res.pca <- PCA(t(pca_snp), graph = FALSE, scale.unit = F)
ind_pca <- cbind(sample_names, res.pca$ind$coord)
rownames(ind_pca) <- NULL

# Percentage of variance explained by each PC
#scree <- fviz_screeplot(res.pca, ncp=10)
# Samples on 2 first PCs
ggplot(ind_pca, aes(x=Dim.1, y=Dim.2, col=Sex, label=Name)) + 
  geom_point() + 
  theme_bw() + 
  geom_text_repel(size=2.5) + 
  ggtitle("PCA after pooling alternative alleles for all SNPs") + 
  xlab("PC1") + 
  ylab("PC2")


```

```{r 3d_interactPCA, eval=F, echo=F}
library("plot3D");library("plot3Drgl")

test_km <- ind_pca
test_km$k_clus <- kmeans(ind_pca[,3:5],centers = 4)$cluster
attach(test_km)
# Create a scatter plot
scatter3D(Dim.1, Dim.2, Dim.3, phi = 10, theta = 80, bty = "b2", pch = 20, cex = 2, type="h", colvar=k_clus)
plotrgl()
# Add text
text3D(x, y, z,  labels = rownames(iris),
        add = TRUE, colkey = FALSE, cex = 0.5)
detach(test_km)

```


# Analysis of highly diverse regions

To know if the alleles in highly diverse regions (Figure \ref{fig:peaks}) segregate according to sample relatedness from the whole genome or if there are significant discordances (e.g. always similar among females), I made a quick visualisation (Figure \ref{fig:zoom_MDS}). It seems the clustering is highly different around the peak on chromosome 4, with females and males clustering together except for a few males.

I can think of two other things to do, not sure if it makes sense. A possible method (discuss with Tanja) is to compute a sample-similarity matrix S~g~ for the whole genome and S~r~ at each region from the individual coordinates on the first 3-5 principal components, and then compute the correlation score between the S~g~ and each of the S~r~. Sum of absolute differences could be used instead of correlation.

Based on the expectation that sex determining regions should be het in females and hom in males, it would also be possible to compute a cluster separation score for haplotypes of females (i.e. intra female vs inter female distance) and to see if all females have more intra divergence in some regions.


```{r cache=T, dependson="Hamming_MDS", div_peaks, echo=F, fig.cap="\\label{fig:peaks}Hex-binned 100bp windows of PI diversity values. Color shows the number of windows falling in the hexbin. Peaks considered afterwards are those where the window PI value is >=0.35 (on chromosomes 3, 4 and 5)."}

PIpeaks <- subset(div100, div100$X4>0.35)

ggplot(div100, aes(x=(X2+X3)/2, y=X4)) + 
  facet_wrap(~X1) + 
  geom_hex()
```

```{r peak_diff, cache=T, dependson="Hamming_MDS", echo=F, fig.height = 9, fig.width = 6, fig.cap="\\label{fig:zoom_MDS}MDS plot of the individual regions (+/- 10kb around) showing similarity between individuals."}

region_size <- 10000; chr <- ""; pk_coord <- 0
peakplots <- list()
gg_idx <- 1


for(peak in 1:nrow(PIpeaks)){
  
  # Skip peaks less than 10kb apart
  if(pull(PIpeaks[peak,"X1"]) == chr & 
     abs(pull(PIpeaks[peak,"X2"]) - pk_coord) < 10000) {next}
  
  chr <- pull(PIpeaks[peak,"X1"])
  pk_coord <- (pull(PIpeaks[peak,"X2"]) + pull(PIpeaks[peak,"X2"])) / 2
  start <- pk_coord-region_size
  end <- pk_coord+region_size
  peak_snp <- snp %>% filter(X1==chr) %>% filter(X2 >= start & X2 <=end)
  mds_peak <- as.matrix(peak_snp[,-c(1,2)])
  mds_peak <- mds_peak[complete.cases(mds_peak),]
  H <- hamming(mds_peak)
  H_MDS <- cmdscale(d=H,k=2)
  colnames(H_MDS) <- c("Dim1","Dim2"); rownames(H_MDS) <- NULL
  H_MDS <- cbind(H_MDS, sample_names)

  peakplots[[gg_idx]] <- ggplot(H_MDS, aes_string(x="Dim1", y="Dim2", col="Sex", label="Name")) + 
    geom_point() + 
    theme_bw() + 
    geom_text_repel(size=2.5) + 
    ggtitle(sprintf("MDS, %s:%i", chr,pk_coord)) + 
    guides(col=F)
  gg_idx <- gg_idx + 1
}

grid.arrange(grobs=peakplots,ncol=2)
```