\documentclass[11pt,a4paper]{report}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage{setspace}
\usepackage{url}
%\usepackage{cite}
\usepackage{fancyhdr}
\usepackage{titlesec}
\usepackage{gensymb}
\titleformat{\subsection}{\itshape\normalsize}{\thesubsection}{1em}{}
% interligne
\usepackage{pdflscape}
\usepackage{placeins}
\usepackage{cite}
\usepackage[left=2cm,
			right=2cm,
			top=3cm,
			bottom=3cm]{geometry}
\author{Cyril Matthey-Doret}
\usepackage{outline} \usepackage{pmgraph} \usepackage[normalem]{ulem}
\usepackage{verbatim}
\pagestyle{fancy}
\setlength{\footskip}{80pt}
% chargement des figures
\usepackage{graphicx}					
\usepackage{wrapfig}
\usepackage[export]{adjustbox}
\usepackage[labelfont=bf]{caption}

\title{
\includegraphics[width=1.75in]{lo_unil06_bleu.pdf} \\
\vspace*{1in}
\textbf{Genetics of sex determination in a parasitoid wasp}}

\author{\Large{Master Project}\\
		Master in Molecular Life Sciences, Bioinformatics\\
				\vspace*{0.5in} \\
		Cyril Matthey-Doret\\
        Supervised by: Casper Van der Kooi\\
        Directed by: Tanja Schwander\\
		\vspace*{0.5in} \\
		Department of Ecology and Evolution\\
        \textbf{University of Lausanne - Switzerland}\\
       } \date{\today}
%--------------------Make usable space all of page
\setlength{\oddsidemargin}{0in} \setlength{\evensidemargin}{0in}
\setlength{\topmargin}{0in}     \setlength{\headsep}{+0.5in}
\setlength{\textwidth}{6.5in}   \setlength{\textheight}{8.5in}
%--------------------Indention
\setlength{\parindent}{1cm}

\begin{document}

%--------------------Title Page
\renewcommand{\headrulewidth}{1pt}
\fancyhead[R]{Master Project, University of Lausanne - Switzerland, December 2017}
\maketitle
%--------------------Begin Outline

\section*{Abstract}
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

\section*{Introduction}
Sexual reproduction is abundant in nature despite the many costs associated with sex, such as the time and energy spent into mate finding and copulation or the risk of disease transmission. This apparent paradox has suscited the interest of biologists for decades and is still a major topic in evolutionary biology. To understand how the benefits of sex can make up for its costs, it is necessary to understand not only the ecology and population dynamics associated with sexual reproduction, but also the underlying genetics that lead to the formation of individuals with different sexes. These genetics mechanisms can have major genome-wide consequences and strong evolutionary implications. A variety of sex determination systems have been identified, they can be either environmental, genetic or a combination of both. To better understand why the strategies to generate different sexes are so diverse and widespread, it is interesting to study when sex is lost. There are indeed asexual species and even species containing both sexual and asexual lineages. Such cases are frequent in Hymenoptera, a megadiverse order comprising 230,000 described species with many different sex determination mechanisms. Among those mechanisms, haplodiploidy is by far the most common. In haplodiploid species, females lay eggs which develop into haploid males if left unfertilized, and into diploid females if fertilized by male sperm. \\
Although haploidy is a well described phenomenon, the genetics underlying the detection of haplodiploidy by the embryo are still very little studied. A particular mechanism named complementary sex determination (CSD) has been described and is known to occur in more than 60 hymenopteran species. In CSD species, sex is determined by only one or a few loci. If two different alleles are present in the organism for at least one CSD locus, the organism develops into a female. Haploid eggs therefore develop into males as they are hemizygous for all loci. In wild populations, on the other hand, there is usually a large allelic diversity at CSD, which usually result in heterozygous CSD loci triggering female development. However, when populations are artificially inbred in laboratory strains, the low allelic diversity will eventually result in diploid individuals with all CSD loci homozygous. Such individuals will develop into abnormal diploid males, which, depending on the species, are not viable, sterile or have reduced fitness. The presence of such diploid males is typically how the presence of CSD is inferred in a species and the frequency of these males has been used in breeding experiments to determine the number of CSD loci in a species from mathematical models.\\
\textit{Lysiphlebus fabarum} is a haplodiploid parasitoid wasp with reproductive polymorphism (both sexual and asexual lineages) where CSD is known to be the mechanism underlying sex determination. As it has both reproductive polymorphism and a simple sex determination system, it has great potential as a model species for the study of questions relating to sex determination and the evolutionary maintenance of sex. In this study, we investigate the genetic basis of CSD in this species. The genetics of CSD is only known in the cape honeybee (\textit{Apis mellifera capensis}) where the \textit{csd} gene was identified to stem from a duplication of the ancestral \textit{fem}. The \textit{csd} gene underwent strong positive selection which resulted in neo functionarlization and it now acts as an RNA splicer of its paralog, \textit{fem}


\section*{Results}

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

\section*{Figures, tables and legends}

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

\FloatBarrier

\section*{Discussion}

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

\section*{Materials and methods}

\subsection*{Crossing experiments}
\textit{Performed prior to master project.}\\
A haploid male coming from an asexual family of \textit{Lysiphlebus fabarum} was crossed with an inbred sexual line. The offspring of asexual females at the 4th generation were used.

\subsection*{RAD-seq protocol}
\textit{Performed prior to master project.}\\
N samples of 4th generation \textit{Lysiphlebus fabarum } coming from X different asexual females kept in ethanol at -XX\degree C for XX months. They were sexed visually and prepared in 6 separate libraries, all following the same protocol from XXX. ddRAD-seq was performed on all samples using EcorI and MseI restriction enzymes and a fragments  of 200-450 bp were size selected on agarose gel. Single-end sequencing was performed using Illumina Miseq (or Hiseq 2500???). Samples were multiplexed in each library following the TruSeq multiplexing design and libraries were pooled pairwise on the same Illumina lane using different adapters (iA06 or iA12).

\subsection*{STACKS pipeline}
We use the STACKS software (version 1.30) \textbf{CITATION} to process RAD-seq data as it is versatile, lends itself well to the analysis of non-model species and is well documented.
Following quality control using fastqc (version 0.11), the raw reads were trimmed and demultiplexed using the "process radtags" module from the STACKS suite and two mismatches were allowed to detect adapters. The 96bp demultiplexed reads were mapped to the latest assembly of the \textit{L. fabarum} genome using BWA aln (version 0.7.2) with 4 mismatches allowed. Only uniquely mapped alignments were extracted using samtools (version 1.3). Stacks were generated from SAM files of unique hits using the Pstacks module, requiring a minimum stack depth (-m) of 3. The catalogue of loci was built with Cstacks allowing for a distance (-n) of 3 mismatches between samples at each locus. Individuals with less than 10\% total radtags compared to the average across all samples were excluded from the analysis. Populations was run pooling all samples together, requiring each locus to be present (-r) in at least 80\% of samples. The different STACKS parameters were selected following guidelines in \textbf{Paris et al. 2017}.

\subsection*{Ploidy separation}
Genome wide heteroyzgosity per individual was computed on all variant sites using the output VCF file from a first populations run with more stringent parameters. Only high confidence loci were included by requiring a minimum sequencing depth (-m) of 20 reads in populations, to minimize chances of assigning the wrong ploidy to any individual. A conservative threshold of 77\% heterozygosity among variant sites was determined empirically \textbf{(Figure SX)} and individuals above that threshold were considered haploid and excluded from the main populations run used the other analyses.

\subsection*{Categorizing families}
The proportion of males among diploid offspring was used to group families by number of heterozygous CSD loci in the mother. The total number of diploid males among non-sequenced individuals was inferred from sequenced individuals by extrapolating the rate of haploidy from sequenced individuals to the total number of males as follows:
\begin{center}
$N_{DM}=N_M*\frac{n_{dm}}{n_{dm}+n_{hm}}$\\
\end{center}
Where upper-case letters represent total individuals count in each family and lower-case letters represent sequenced individuals count in the family. $N_M$ is the number of males, $N_{HM}$ is the number of haploid males and $N_{DM}$ the number of diploid males.

The proportion of males among diploid offspring was then computed using the inferred number of diploid males. The families were classified using 1-dimensional k-means clustering on the proportion of males among diploid offspring and 2 scenarios were considered to decide the number of categories of families; either 2 CSD loci, resulting in 3 categories (k=3), or 3 CSD loci resulting in 7 categories (k=7).\\

\subsection*{Finding centromeres}
All fixed and variant sites were extracted using the --genomic parameter of populations. For families where the mother was available, we excluded all sites that were either homozygous or missing in the mother. For other families, we excluded sites that were missing of homozygous in all offspring in the family. 
The proportion of heterozygous sites was computed along the chromosomes in a sliding window containing 30 sites. In each chromosome, the centromere was considered to be in the window with the lowest proportion of heterozygous sites.

\subsection*{Terminal or Central fusion automixis}
Lysiphlebus is 

\subsection*{Association mapping}
Case-control association mapping was carried separately in each category of family. The number of heterozygous males, heterozygous females, homozygous males and homozygous females was computed for every SNP and a two-sided Fisher exact-test was performed on the 2X2 contingency tables. P-values were corrected for multiple-testing using Benjamini-Hochberg correction. 

\section*{Supplementary Material}

Code and manuscript data hosted at: [INSERT URLS]

%\bibliography{thesis}{}
\bibliographystyle{apalike}
\fancyhead[L]{\slshape }
\bibliography{thesis}
\end{document}